src/chainstate/stacks/auth.rs

```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+    pub fn make_sighash(cur_sighash: &Txid, cond_code: &TransactionAuthFlags, pubkey: &StacksPublicKey, sig: &MessageSignature) -> Txid {
+        // new hash combines the previous hash and all the new data this signature will add.  This
+        // includes:
+        // * the previous hash
+        // * the auth flag
+        // * the public key's compressed bit
+        // * the public key
+        // * the signature
+        let new_tx_hash_bits_len = 32 + 1 + 1 + STACKS_PUBLIC_KEY_ENCODED_SIZE + MESSAGE_SIGNATURE_ENCODED_SIZE;
+        let mut new_tx_hash_bits = Vec::with_capacity(new_tx_hash_bits_len as usize);
+        let pubkey_encoding = 
+            if pubkey.compressed() {
+                TransactionPublicKeyEncoding::Compressed
+            }
+            else {
+                TransactionPublicKeyEncoding::Uncompressed
+            };
+
+        let pubkey_buf = StacksPublicKeyBuffer::from_public_key(pubkey);
+
+        new_tx_hash_bits.extend_from_slice(cur_sighash.as_bytes());
+        new_tx_hash_bits.extend_from_slice(&[*cond_code as u8]);
+        new_tx_hash_bits.extend_from_slice(&[pubkey_encoding as u8]);
+        new_tx_hash_bits.extend_from_slice(pubkey_buf.as_bytes());
+        new_tx_hash_bits.extend_from_slice(sig.as_bytes());
+
+        assert!(new_tx_hash_bits.len() == new_tx_hash_bits_len as usize);
+
+        let next_sighash = Txid::from_sighash_bytes(&new_tx_hash_bits);
+        next_sighash
+    }
+
+    /// Linear-complexity signing algorithm -- we sign a rolling hash over all data committed to by
+    /// the previous signer (instead of naively re-serializing the transaction each time).
+    /// Calculates and returns the next signature and sighash, which the subsequent private key
+    /// must sign.
+    pub fn next_signature(cur_sighash: &Txid, cond_code: &TransactionAuthFlags, privk: &StacksPrivateKey) -> Result<(MessageSignature, Txid), net_error> {
+        // sign the current hash
+        let mut digest_bits = [0u8; 32];
+        digest_bits.copy_from_slice(cur_sighash.as_bytes());
+
+        let sig = privk.sign(&digest_bits)
+            .map_err(|se| net_error::SigningError(se.to_string()))?;
+
+        let pubk = StacksPublicKey::from_private(privk);
+        let next_sighash = TransactionSpendingCondition::make_sighash(cur_sighash, cond_code, &pubk, &sig);
+        
+        Ok((sig, next_sighash))
+    }
+    
+    /// Linear-complexity verifying algorithm -- we verify a rolling hash over all data committed
+    /// to by order of signers (instead of re-serializing the tranasction each time).
+    /// Calculates the next sighash and public key, which the next verifier must verify.
+    /// Used by StacksTransaction::verify*
+    pub fn next_verification(cur_sighash: &Txid, cond_code: &TransactionAuthFlags, key_encoding: &TransactionPublicKeyEncoding, sig: &MessageSignature) -> Result<(StacksPublicKey, Txid), net_error> {
```

Comment:
Where does this function verify the current signature? I see digest_bits being filled, but they don't appear to get checked.

Reply:
Ah, so this algorithm is kind of interesting.  The top-level verification methods you want to look at are `SinglesigSpendingCondition::verify()` and `MultisigSpendingCondition::verify()`, which check the output of this method.

The call to the `recover_to_pubkey` method works like an ECDSA signature verification in that it takes the _(r, s)_ signature point, the plaintext digest, and a 1-byte "recovery code," and turns it back into a public key (the recovery code indicates _which_ of a set of up to two recovered public keys to use).  At the very end of verifying the transaction in `SinglesigSpendingCondition::verify()` and `MultisigSpendingCondition::verify()`, the public key(s) are hashed and compared to the address of the spending condition, which completes the check.  Note that using a different plaintext digest for the same _(r, s)_ signature will produce a _different_ public key, meaning that if the plaintext is modified, then the resulting public key hash will not match the address.

Reply:
I see -- so the signature verification is happening in `StacksPublicKey::recover_to_pubkey` -- in that case, the `digest_bits` variable is an unused variable?

Reply:
Yup; will remove.

Reply:
Fixed in d399b6d9